pulsar-io/couchbase/src/main/java/org/apache/pulsar/io/cb/CBAbstractSource.java

```diff
@@ -0,0 +1,314 @@
+            id = DcpMutationMessage.bySeqno(event);
+          }
+          else if (DcpDeletionMessage.is(event)) {
+            message = MessageUtil.getContentAsByteArray(event);
+            id = DcpDeletionMessage.bySeqno(event);
+          }
+        }
+        else {
+          message = new byte[event.readableBytes()];
+          event.readBytes(message);
+        }
+
+        try {
+          CBRecord cbRecord = new CBRecord();
+          cbRecord.setRecord(extractValue(message));
+          cbRecord.setId(id);
+          consume(cbRecord);
+
+          flowController.ack(event);
+          event.release();
+        }
+        catch (Exception ex) {
+          log.error("process error!", ex);
+
+          event.retain();
+        }
+      }
+    });
+
+    client.controlEventHandler(new ControlEventHandler() {
+      @Override
+      public void onEvent(ChannelFlowController flowController, ByteBuf event) {
+        try {
+          log.info("Control Event: {}", event.toString());
+
+          if (DcpSnapshotMarkerRequest.is(event) && cbSourceConfig.getUseSnapshots()) {
+            DcpSnapshotMarkerRequestMeta snapshot = new DcpSnapshotMarkerRequestMeta(event);
+
+            DcpSnapshotMarkerRequestMeta existingSnapshot = snapshots.put(snapshot.getPartition(), snapshot);
+
+            if (existingSnapshot != null) {
+              log.warn("Incomplete snapshot detected: {}", existingSnapshot);
+            }
+          }
+
+          if (RollbackMessage.is(event)) {
+            final short vbucket = RollbackMessage.vbucket(event);
+            final long seqNo = RollbackMessage.seqno(event);
+
+            log.warn("Rolling back partition {} to seqno {}", vbucket, seqNo);
+
+            // Netty IO thread
+            client.rollbackAndRestartStream(vbucket, seqNo)
+                .subscribe(new CompletableSubscriber() {
+                  private Subscription subscription;
+
+                  @Override
+                  public void onCompleted() {
+                    log.info("Rollback for vbucket {} complete", vbucket);
+
+                    // TODO: find if exclusive unsubscribe needed ?
+                    if (!subscription.isUnsubscribed()) {
+                      subscription.unsubscribe();
+                    }
+                  }
+
+                  @Override
+                  public void onError(Throwable throwable) {
+                    log.error("Failed to rollback vbucket {} to seqNo {}", vbucket, seqNo, throwable);
+                  }
+
+                  @Override
+                  public void onSubscribe(Subscription subscription) {
+                    log.info(subscription.toString());
+
+                    this.subscription = subscription;
+                  }
+                });
+          }
+        }
+        catch (Throwable throwable) {
+          log.error("Exception in ControlEventHandler", throwable);
+        }
+        finally {
+          //TODO: push to queue
```

Comment:
Is it still necessary to improve TODO?

Reply:
`ack` and `release`? how it is going on when `onEvent` encounter an `Exception`? use `BlockingQueue` to offet it ?